### DISCUSSION TITLE
RustDesk Installation: Certificate & Nginx Error

### DISCUSSION THREAD

**Discussion by isential-gmbh** [rustdesk/rustdesk-server-pro#220] [Q&A]
I had reached out to you when I attempted to install RustDesk a few weeks ago. Here's the link if you're interested: https://github.com/rustdesk/rustdesk-server-pro/discussions/209

After being out due to illness, I'm attempting it again. However, this time I have a subdomain for RustDesk—let's call it rd.example.tld.

I'm attempting to install RustDesk using the "Simple Install Script." During the installation process, I'm prompted to install Certbot. However, since Certbot is already installed (via Plesk), and the domain for RustDesk is secured with a Let's Encrypt certificate, I opt not to install Certbot by selecting "No" for the prompt.

Despite my choice, I encounter the following error message:

"Sorry, the TLS certificate for rd.example.tld failed to generate! Please verify that ports 80 and 443 are correctly forwarded and that the DNS record points to this server's IP address. Please try again."

After this error, the script terminates, and I observe the following output in the console:

"Error while running nginx -c /etc/nginx/nginx.conf -t.

nginx: [emerg] unknown directive "brotli" in /etc/nginx/conf.d/brotli.conf:1 nginx: configuration file /etc/nginx/nginx.conf test failed

The nginx plugin is not functioning correctly; there may be issues with the existing configuration.

The error reported is: MisconfigurationError('Error while running nginx -c /etc/nginx/nginx.conf -t.\n\nnginx: [emerg] unknown directive "brotli" in /etc/nginx/conf.d/brotli.conf:1\nnginx: configuration file /etc/nginx/nginx.conf test failed\n')"

Could you please assist me in identifying the problem and suggesting a solution?

**Comment by rustdesk** [comment]
Could you consider docker installation, https://rustdesk.com/docs/en/self-host/rustdesk-server-pro/installscript/docker/?

**Comment by rustdesk** [comment]
BTW, I think your nginx issue has nothing to do with the insintall script, since the script has never used `brotli` etc.

Since nobody maintains the installation script any more, I will remove it from the manual later, and only recommend docker installation.

**Reply by ZetaWaves** [reply]
I would hope you're not going to get rid of the installation script. Many people use that to install, upgrade, backup, restore. It's very important to much of the community. Docker is extremely buggy with RustDesk, but I understand why many people use it. Most of the guides and how-to's on your website are directed to people who have Linux installations, not users who use docker.

**Reply by rustdesk** [reply]
Thanks

**Comment by ZetaWaves** [comment]
> I had reached out to you when I attempted to install RustDesk a few weeks ago. Here's the link if you're interested: #209
> 
> After being out due to illness, I'm attempting it again. However, this time I have a subdomain for RustDesk—let's call it rd.example.tld.
> 
> I'm attempting to install RustDesk using the "Simple Install Script." During the installation process, I'm prompted to install Certbot. However, since Certbot is already installed (via Plesk), and the domain for RustDesk is secured with a Let's Encrypt certificate, I opt not to install Certbot by selecting "No" for the prompt.
> 
> Despite my choice, I encounter the following error message:
> 
> "Sorry, the TLS certificate for rd.example.tld failed to generate! Please verify that ports 80 and 443 are correctly forwarded and that the DNS record points to this server's IP address. Please try again."
> 
> After this error, the script terminates, and I observe the following output in the console:
> 
> "Error while running nginx -c /etc/nginx/nginx.conf -t.
> 
> nginx: [emerg] unknown directive "brotli" in /etc/nginx/conf.d/brotli.conf:1 nginx: configuration file /etc/nginx/nginx.conf test failed
> 
> The nginx plugin is not functioning correctly; there may be issues with the existing configuration.
> 
> The error reported is: MisconfigurationError('Error while running nginx -c /etc/nginx/nginx.conf -t.\n\nnginx: [emerg] unknown directive "brotli" in /etc/nginx/conf.d/brotli.conf:1\nnginx: configuration file /etc/nginx/nginx.conf test failed\n')"
> 
> Could you please assist me in identifying the problem and suggesting a solution?

You need to run the install and choose yes. Choosing no will drop into the other if statement and will mess up your installation temporarily. Re-run the script and choose yes. Verify symlinks /snap snap are created/active. Verify snap and certbot are installed.

**Reply by isential-gmbh** [reply]
I did that and recorded it at the same time so you can see what happened. Here's the video:

https://isential.de/tmp/rd-installation.mp4

I could try to perform the installation on Docker, but now I'm uncertain. You mentioned that RustDesk is quite error-prone under Docker. Would you really advise against using Docker?"

**Reply by ZetaWaves** [reply]
sudo ln -s /var/lib/snapd/snap /snap

Verify snap is installed. Verify certbot is installed. Verify nginx is installed. Remove/uninstall brotli.

Are you running in CentOS? Or AlmaLinux 8?

**Reply by isential-gmbh** [reply]
I repeated the installation procedure, but this time I directly provided the IP address. There was no issue with this, and the installation was successfully completed.

Regarding your question about the operating system: I am using Ubuntu 22.04.

I assume that I made the mistake of specifying a real subdomain 'sub.example.tld', which led to the problem with nginx. I should have simply created a DNS A record for 'sub.example.tld' pointing to the IP address. At least, that's what I suspect. What do you think? Am I correct in this assumption?

I am now trying to do what I intended to do more than a week ago... I will let you know if I encounter any problems again.

**Reply by ZetaWaves** [reply]
Yes, sounds like you had a DNS issue then.

Public domain should always be pointed to the IP address of the server. The server itself should point to localhost or the public IP address of the server. localhost would be best in case you ever change your public IP.

**Reply by isential-gmbh** [reply]
I don't understand your statement '_Public domain should always be pointed to the IP address of the server. The server itself should point to localhost or the public IP address of the server. localhost would be best in case you ever change your public IP._'. What does that mean?

One more question: When installing RustDesk, I specify the public IP address of the RustDesk server. Do I need to provide this IP address when setting up the clients, or can I create an A record on another server that points to the public IP address of my RustDesk server and use that A record instead?
